src/event/event_queue.rs

```diff
@@ -64,30 +65,48 @@ impl EventQueue {
     /// Creates an unbounded queue for scheduling closures on the JavaScript
     /// main thread
     pub fn new<'a, C: Context<'a>>(cx: &mut C) -> Self {
-        let tsfn = unsafe { ThreadsafeFunction::new(cx.env().to_raw(), Self::callback) };
+        #[cfg(feature = "napi-6")]
+        let trampoline = InstanceData::threadsafe_trampoline(cx);
 
-        Self {
-            tsfn,
-            has_ref: true,
-        }
+        #[cfg(not(feature = "napi-6"))]
+        let trampoline = ThreadsafeTrampoline::new(env);
+
+        let mut queue = Self {
+            trampoline: trampoline,
+            has_ref: false,
+        };
+
+        // Start referenced
+        queue.reference(cx);
+
+        queue
     }
 
     /// Allow the Node event loop to exit while this `EventQueue` exists.
     /// _Idempotent_
     pub fn unref<'a, C: Context<'a>>(&mut self, cx: &mut C) -> &mut Self {
```

Comment:
`unref` will not operate correctly in the presence of mixed usage. It will immediately `unref` the shared queue even if referenced copies still exist.

Reply:
I think it does the checks here to make sure that global instance is referenced/unreferenced properly. I'll move the logic into the `trampoline.rs`.

Reply:
I believe I've addressed this by making `reference()`/`unref()` not idempotent in `ThreadsafeTrampoline` in the latest commit in this PR. The way it works is that we enforce idempotency on the level of `EventQueue` and only when the value of `has_ref` actually changes - we call `reference()`/`unref()` on the parent `ThreadsafeTrampoline`. It in turn will track the total number of references and invoke `ref_threadsafe_function`/`unref_threadsafe_function` whenever the internal reference count becomes zero or non-zero.